src/Test.UnitTests.Sarif.WorkItemFiling/Filtering/AllResultsFilteringStrategyTests.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using FluentAssertions;
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.WorkItemFiling.Filtering;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Test.UnitTests.Sarif.WorkItemFiling.Filtering
+{
+    public class AllResultsFilteringStrategyTests
```

Comment:
>AllResultsFilteringStrategyTests [](start = 17, length = 32)

Can I suggest that you please create some early negative test cases? It is important to understand API requirements in exceptional cases, for example. i.e., what does the code look like to handle a case where one of our grouping/refactoring strategies throws an exception? something I think is important to drive into the design early. I usually author concrete types that can be initialized via ctor with data that describes what behaviors they should engage in. In addition to unhandled exceptions, we should be considering some eventual machinery around hang conditions/time out. This is tricky to implement, as you know. We will likely need to address this at some point but perhaps not right away. We certainly should tackle the problem of exceptions in absolutely every possible component/layer in the design from the beginning, imo.

Reply:
This comment actually applies to `WorkItemFilerTests`, where the question is "What happens to the filer when a strategy throws an exception?". I can easily test that by mocking up strategy objects that throw. The answer, of course, as the code stands, is "The exception escapes from the `WorkItemFiler`", and I'll write the tests that say that.

As for this class (`AllResultsFilteringStrategy`) we do have a test for the only exception it explicitly throws (`ArgumentNullException`).

Working on this now..

---
In reply to: [306121895](https://github.com/microsoft/sarif-sdk/pull/1582#discussion_r306121895) [](ancestors = 306121895)

Reply:
Fixed.

---
In reply to: [306355604](https://github.com/microsoft/sarif-sdk/pull/1582#discussion_r306355604) [](ancestors = 306355604,306121895)

Reply:
My larger point is that unhandled exceptions in all possible layers/components s/be explicitly tested before we declare victory. For asynchronous API, exception handling is critical, so that exceptions don't get swallowed/unreported/cause hangs when they occur.

---
In reply to: [306365826](https://github.com/microsoft/sarif-sdk/pull/1582#discussion_r306365826) [](ancestors = 306365826,306355604,306121895)